### PR TITLE
PPM handset support

### PR DESCRIPTION
**Pull request by pkendall64**
Extracts the serial handling code to a `Handset` class with three implementations

1. CRSFHandset
2. PPMHandset
3. AutoDetect 

The `AutoDetect` controller checks the input signal and if it detects a UART then replaces the handset implementation with the CRSF and it it detects a PPM signal then the PPM handset is used.

### CODE REVIEW

**Comment by YI-BOYANG** [comment]
It sound good.

**Review by JyeSmith — APPROVED** [review]
This is a huge PR and has been tested as working.  Itll get a heap more testing during the 3.4 testing in the coming weeks.

**Comment by pkendall64** [comment]
> This is a huge PR and has been tested as working. It'll get a heap more testing during the 3.4 testing in the coming weeks.

Yeah it is rather large and should have plenty of testing.

File: src/lib/Controller/AutoDetect.cpp [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+void AutoDetect::startCRSF()
+{
+    Controller *crsf = new CRSFController();
+    crsf->setRCDataCallback(RCdataCallback);
+    crsf->registerParameterUpdateCallback(RecvParameterUpdate);
+    crsf->registerCallbacks(connected, disconnected, RecvModelUpdate);
+    crsf->Begin();
+    crsf->setPacketInterval(RequestedRCpacketInterval);
+    delete this;
+    controller = crsf;
```
Comment (CapnBry) [inline src/lib/Controller/AutoDetect.cpp:63]: The code between PPM and CRSF is identical apart from the `new`, refactor to use a common function to do all the stuff?
```cpp
void AutoDetect::migrateTo(Controller *that)
{
    that->setRCDataCallback(RCdataCallback);
    that->registerParameterUpdateCallback(RecvParameterUpdate);
    that->registerCallbacks(connected, disconnected, RecvModelUpdate);
    that->Begin();
    that->setPacketInterval(RequestedRCpacketInterval);
    delete this;
    controller = that;
}

void AutoDetect::startCRSF()
{
    migrateTo(new CRSFController()):
}
```

**Review by CapnBry — COMMENTED** [review]
Works great in all the test cases I could throw at it-- using trainer port and 4, 5,6 channels, then just slapping it in the module bay.

Clever solutions to the static CRSF stuff as well as supporting IsArmed if there aren't enough channels. 😺

File: src/lib/Controller/PPMController.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#include "controller.h"
+
+#include <driver/rmt.h>
```
[the remark this answers is not in the document]
Comment (pkendall64) [reply src/lib/Controller/PPMController.h:3]: Required by type used in the private section. `RingbufHandle_t`

File: src/lib/Controller/AutoDetect.cpp (hunk 1 above)
Comment (pkendall64) [reply src/lib/Controller/AutoDetect.cpp]: Done

File: src/lib/Controller/AutoDetect.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+void AutoDetect::handleInput()
+{
+    size_t length = 0;
+    auto now = millis();
+    static auto lastDetect = 0LU;
```
[the remark this answers is not in the document]
Comment (pkendall64) [reply src/lib/Controller/AutoDetect.cpp]: Done

File: src/lib/Controller/controller.h [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ *
+ * There are three implementations of the Controller class
+ *
+ * - CRSFController - implements the CRSF protocol for communicating with the handset
+ * - PPM Controller - PPM protocol, can be connected to the DSC/trainer port for simple non-CRSF handsets
+ * - AutoDetect - this implementation uses an RMT channel to auto-detect a PPM or CRSF handset and swap the
+ *   global controller variable to point to instance of the actual implementation. This allows a TX module
+ *   to be moved between a CRSF capable handset and PPM only handset e.g. an EdgeTX radio and a surface radio.
+ */
+class Controller
```
[the remark this answers is not in the document]
Comment (pkendall64) [reply src/lib/Controller/controller.h]: Well, I renamed it `Handset`... just because it's you ❤️

**Review by CapnBry — APPROVED** [review]
Looks good, excellent work. Thanks for the rename too, I really appreciate it :)

**Comment by jmptbl** [comment]
Am I understanding correctly that this update will allow old PPM transmitters to get ELRS support with any ELRS TX module?

How would one configure all of the ELRS options in the TX module without the OpenTX/LUA environment?

**Comment by pkendall64** [comment]
> Am I understanding correctly that this update will allow old PPM transmitters to get ELRS support with any ELRS TX module?
> 
> How would one configure all of the ELRS options in the TX module without the OpenTX/LUA environment?

You are correct. You would need a module with an OLED and joystick to be able to configure the settings.
Later, we may make it that the settings can be configured in the web-UI.

**Comment by jmptbl** [comment]
> You are correct. You would need a module with an OLED and joystick to be able to configure the settings. Later, we may make it that the settings can be configured in the web-UI.

I am literally on the edge of adding ELRS to a Multiplex Royal Evo, and I was planning to use a PPM-ELRS converter from AliExpress. This is so much better!  Really, thank you so much for this. I can't wait to try it!

**Comment by jmptbl** [comment]
@pkendall64 My Multiplex Royal Evo now speaks ExpressLRS, thanks to your work.  🥳🙏🏻

**Comment by hz89377361** [comment]
The R9M can't use PPM, the self-test fails, the PPM signal can't be detected, the remote control is JR 11X 8CH PPM, and also tried the default PPM settings of the X9E, it can't be detected either, the library file version is https://github.com/pkendall64/ExpressLRS.git

**Comment by pkendall64** [comment]
PPM Handset detection is not supported on STM based TX modules. Only ESP32.

**Comment by jmptbl** [comment]
Hi @pkendall64 , when using PPM are there any benefits in using packet rates above 50 Hz?  Since PPM only runs at around 40-50 Hz.

**Comment by pkendall64** [comment]
Only in so far as you'll have more chances to get the data to the receiver. On EdgeTX I know you can adjust the PPM frame timing, but you'd still no be any better than about 90Hz. So it's probably better to stick to 50Hz and use it for the range.

**Comment by hgf1989** [comment]
ok,i know ,ppm should set 50hz at elrs , but how to connect . use the same port? crsf? need to do anything?

**Comment by pkendall64** [comment]
> ok,i know ,ppm should set 50hz at elrs , but how to connect . use the same port? crsf? need to do anything?

No there is nothing to do. Just supply a PPM signal to the S.PORT pin on the JR connector, and a common GND and it should work as expected.

**Comment by hgf1989** [comment]
> > ok,i know ,ppm should set 50hz at elrs , but how to connect . use the same port? crsf? need to do anything?
> 
> No there is nothing to do. Just supply a PPM signal to the S.PORT pin on the JR connector, and a common GND and it should work as expected.

AutoDetect ？

**Comment by pkendall64** [comment]
> AutoDetect ？

Yes, it will auto-detect a PPM or serial signal on the S.PORT pin

**Comment by NoobNinjaLemon** [comment]
Noice

**Comment by xros** [comment]
Is the PPM protocol only supported on the TX? Can the RX output in PPM?

There are some old FC board only supports PPM. If you can, please add support for RX as well. Appreciate your work!

**Comment by pkendall64** [comment]
Yes it is only supported on TX modules to support handsets that don't support CRSF protocol.

I don't think we'd add PPM output on receivers because it would not work very well at all on the 8285 based receivers as they just don't have the hardware support for it and we'd have to use the timer that we use for the PWM output. On these receivers the output is already very jittery. ESP32 based receivers could easily do it because it has the RMT device which would be perfect for this, but I just don't it's worth the effort.

**Comment by wimalopaan** [comment]
Would this also work on RX-as-TX?

**Comment by pkendall64** [comment]
> Would this also work on RX-as-TX?

Not on ESP8285 receivers, as it utilises the RMT module in the ESP32.

**Comment by Harosteadi** [comment]
Hello, my idea is to control a camera gimbal with a HID device (mouse). I will translate that HID signal into PPM following [this project](https://github.com/Doctor-Proton/USB_to_RC) and that PPM signal to a TX module like [this one](https://es.aliexpress.com/item/1005005257248373.html) now that I know ELRS transmitters can read PPM signal not only CRSF. Do you find an easier way to do it? What I am saying is correct? I want to use the TX module without a handunit (controller) it is possible right? Just need the TX module to have screen and menu button, right? Thanks a lot.

**Comment by pkendall64** [comment]
Yes, that should work fine. You don't need a screen on the module, but it will help so you can see the telemetry and adjust the parameters.

**Comment by yonmudi** [comment]
Hello, can someone help me do an experiment on the flysky GT3B ppm (DSC Port) with the TX jumper module aion nano elrs firmware 3.5.0. The problem is that the module always restarts when the throttle/steering is moved. I have also tried it with an external battery module but it still restarts/resets the oled screen.

**Comment by DroneWerks** [comment]
> > Am I understanding correctly that this update will allow old PPM transmitters to get ELRS support with any ELRS TX module?
> > How would one configure all of the ELRS options in the TX module without the OpenTX/LUA environment?
> 
> You are correct. You would need a module with an OLED and joystick to be able to configure the settings. Later, we may make it that the settings can be configured in the web-UI.

I recently put in an order for the radiometer nomad with the intention of using it with my old Futaba T6J. I pulled the trigger after hearing news that PPM handsets are now compatible but I have two questions. 

How would the ELRS module communicate with the screen? Does this require excyra hardware/wiring or just the ppm signal from the trainer port?? 

Also what are the options to install a telemetry beeper/is there a pin on the ELRS module I can use to signal the beeper?

**Comment by pkendall64** [comment]
You would need a module with an OLED screen so you can change the parameters. On the OLED we also have telemetry screen with LQ and RSSI etc.

You can plug the module into a regular radio and set up model 0 as you wish, then move it to a PPM connected handset and it will use those parameters.

**Comment by DroneWerks** [comment]
Understood.

Do you think it would be feasible to falsh a ESP32*oled with lua script and use this to configure the tx module?

If so would i just fasah with crsf and  communicate through the crsf 3 pin connector?

Or am i just making this too complicated and will be able to adjust configuration through the web interface?

FYI i dont have an transmitter with edge or open tx just the old futaba

**Comment by pkendall64** [comment]
It would be a whole other development to have a configuration via another ESP32 with an OLED.
You can currently configure it via the web UI by downloading the config JSON file and modifying the values then uploading it again. Not the nicest experience but do-able.

**Comment by Divefire** [comment]
Now this has been implemented, how feasible is it to add support for faster (shorter) PPM frames? I had in mind being able to shorten the PPM frame so redundant space (for less ch users) is eliminated, thus reducing latency between tx input and servo output.

Futaba HRS support for 2ch transmitters would also be great but I don't know if that has been reversed engineered or if it's a PPM or PCM base.

**Comment by Divefire** [comment]
> > Now this has been implemented, how feasible is it to add support for faster (shorter) PPM frames? I had in mind being able to shorten the PPM frame so redundant space (for less ch users) is eliminated, thus reducing latency between tx input and servo output.
> 
> It already handles short PPM frames. It will auto-detect channels in the frame, and end the frame when it sees no changes for 4ms.

Thanks for the feedback, that's really interesting to know and opens up a lot of retrofitting use cases. for ELRS. Do you know if it would it be possible to make the 4ms end frame window a variable parameter though? In terms of responsiveness having that as tight as possible would be a boon I think?

**Comment by pkendall64** [comment]
> Thanks for the feedback, that's really interesting to know and opens up a lot of retrofitting use cases. for ELRS. Do you know if it would it be possible to make the 4ms end frame window a variable parameter though? In terms of responsiveness having that as tight as possible would be a boon I think?

Totally possible, heres the line in the code that configures the 4ms end frame detection.
https://github.com/ExpressLRS/ExpressLRS/blob/311f975eb54d3d6180c9a20f08996937c7b43639/src/lib/Handset/PPMHandset.cpp#L20
However for such a rare use case I don't think I'd be happy having it as a configurable parameter. It's easy enough to change the code and rebuild.

**Comment by Divefire** [comment]
> Totally possible, heres the line in the code that configures the 4ms end frame detection.
> 
> https://github.com/ExpressLRS/ExpressLRS/blob/311f975eb54d3d6180c9a20f08996937c7b43639/src/lib/Handset/PPMHandset.cpp#L20
> 
> However for such a rare use case I don't think I'd be happy having it as a configurable parameter. It's easy enough to change the code and rebuild.

Ah I'm with you, thanks for the code link, I'll start building in the new year then and see where the sweet spot lies for me. Appreciate the help!

**Comment by hpcre** [comment]
How would this connect to a diy tx module based on esp32? Where would I connect the ppm output of my transmitter to the esp32 board?

**Comment by pkendall64** [comment]
The normal CRSF pin that you use to connect to a radio. That pin will auto-detect PPM vs Serial CRSF.

**Comment by wimalopaan** [comment]
Is there any hope that this get fixed in the near future?

#3518

**Pull request merged**